python/cuml/decomposition/pca_mg.pyx

```diff
@@ -0,0 +1,346 @@
+                                params.n_rows, params.n_cols, n_part_row)
+
+        cdef uintptr_t comp_ptr = get_dev_array_ptr(self.components_ary)
+
+        cdef uintptr_t explained_var_ptr = \
+            get_cudf_column_ptr(self.explained_variance_)
+
+        cdef uintptr_t explained_var_ratio_ptr = \
+            get_cudf_column_ptr(self.explained_variance_ratio_)
+
+        cdef uintptr_t singular_vals_ptr = \
+            get_cudf_column_ptr(self.singular_values_)
+
+        cdef uintptr_t mean_ptr = get_cudf_column_ptr(self.mean_)
+
+        cdef uintptr_t noise_vars_ptr = \
+            get_cudf_column_ptr(self.noise_variance_)
+
+        cdef uintptr_t t_input_ptr = get_dev_array_ptr(self.trans_input_)
+
+        cdef cumlHandle* handle_ = <cumlHandle*><size_t>self.handle.getHandle()
+
+        cdef uintptr_t data
+        if self.dtype == np.float32:
+            data = self._build_dataFloat(arr_interfaces)
+            
+            fit_transform(handle_[0],
+                <RankSizePair**>rankSizePair,
+                <size_t> n_total_parts,
+                <floatData_t**> data,
+                <float*> t_input_ptr,
+                <float*> comp_ptr,
+                <float*> explained_var_ptr,
+                <float*> explained_var_ratio_ptr,
+                <float*> singular_vals_ptr,
+                <float*> mean_ptr,
+                <float*> noise_vars_ptr,
+                params,
+                True)
+            
+            self.handle.sync()
+            self._freeFloatD(data, arr_interfaces)
+
+        else:
+            data = self._build_dataDouble(arr_interfaces)
+            
+            fit_transform(handle_[0],
+                <RankSizePair**>rankSizePair,
+                <size_t> n_total_parts,
+                <doubleData_t**> data,
+                <double*> t_input_ptr,
+                <double*> comp_ptr,
+                <double*> explained_var_ptr,
+                <double*> explained_var_ratio_ptr,
+                <double*> singular_vals_ptr,
+                <double*> mean_ptr,
+                <double*> noise_vars_ptr,
+                params,
+                True)
+              
+            self.handle.sync()
+            self._freeDoubleD(data, arr_interfaces)
+
+            # make sure the previously scheduled gpu tasks are complete before the
+        # following transfers start
+        
+        for idx in range(n_total_parts):
+            free(<RankSizePair*>rankSizePair[idx])
+        free(<RankSizePair**>rankSizePair)
+
+        # Keeping the additional dataframe components during cuml 0.8.
+        # See github issue #749
+        self.components_ = cudf.DataFrame()
+        for i in range(0, params.n_cols):
+            n_c = params.n_components
+            self.components_[str(i)] = self.components_ary[i*n_c:(i+1)*n_c]
+
+        
+        if (isinstance(X, cudf.DataFrame)):
+            del(X_m)
+        
+        if not _transform:
+            del(self.trans_input_)
```

Comment:
There's no need for this to be an instance variable, either in the MG nor in the SG case. Also, if `fit_transform()` ends up getting called, this never gets removed from the instance, which means the transformed memory cannot be garbage collected by Python and the resulting memory will end up being serialized w/ the model. This can be a significant overhead, since it's on the order of n_training_samples * n_components, which could grow very large. 

This becomes an even bigger issue in the MG case, since only one of the models trained across the cluster is serialized and brought to the client. In this case, it adds the serialization overhead but also makes the payload unnecessarily large as the cost to move a serialized model, which is normally bounded by the number of features & the number of components, now grows to the number of samples in the partition that ended up being serialized and transported to the client. 

When I started helping you with this, I had originally turned this into a temporary variable but it looks like it's been changed back into an instance variable. I would strongly consider making this a temporary variable and only returning it in the case where `_transform==True` so that the `fit_transform()` function can return control over its underlying CUDA memory to the user.

It doesn't look like scikit-learn keeps an instance variable for `trans_input` and I don't see any of the other PCA functions manipulating or returning it either.

Reply:
Still updating this piece. Thanks for the comment!

Reply:
What's really nice about most of these models being able to do embarrassingly parallel transforms/predictions is that the underlying SG model can be serialized / deserialized separately from the Dask model. 

This allows users the flexibility to train an SG model and then use it in an MG enviornment to scale their transforms/predictions.

They can also go the opposite route and train an MNMG model and then load it in an SG environment to do the predictions/transforms.

Reply:
> Still updating this piece. Thanks for the comment!

No worries. I noticed i was tagged as a reviewer so I figured I'd start looking it over.